third_party/move/move-prover/boogie-backend/src/bytecode_translator.rs

```diff
@@ -402,83 +409,234 @@ impl<'env> StructTranslator<'env> {
-                format!(
-                    "${}: {}",
-                    field.get_name().display(env.symbol_pool()),
-                    self.boogie_type_for_struct_field(
-                        &field.get_id(),
-                        env,
-                        &self.inst(&field.get_type())
+        if struct_env.has_variants() {
+            let constructor = struct_env
+                .get_variants()
+                .map(|variant| {
+                    let struct_variant_name =
+                        boogie_struct_variant_name(struct_env, self.type_inst, variant);
+                    let variant_fields = struct_env
+                        .get_fields_of_variant(variant)
+                        .map(|f| {
+                            format!(
+                                "{}: {}",
+                                boogie_field_sel(&f),
+                                self.boogie_type_for_struct_field(
+                                    &f.get_id(),
+                                    env,
+                                    &self.inst(&f.get_type())
+                                )
+                            )
+                        })
+                        .join(", ");
+                    format!("    {}({})", struct_variant_name, variant_fields,)
+                })
+                .join(", \n");
+            emitln!(writer, "{}", constructor);
+            emitln!(writer, "}");
+        } else {
+            let fields = struct_env
+                .get_fields()
+                .map(|f| {
+                    format!(
+                        "{}: {}",
+                        boogie_field_sel(&f),
+                        self.boogie_type_for_struct_field(
+                            &f.get_id(),
+                            env,
+                            &self.inst(&f.get_type())
+                        )
                     )
-                )
-            })
-            .join(", ");
-        emitln!(writer, "    {}({})", struct_name, fields,);
-        emitln!(writer, "}");
+                })
+                .join(", ");
+            emitln!(writer, "    {}({})", struct_name, fields,);
+            emitln!(writer, "}");
+        }
 
         let suffix = boogie_type_suffix_for_struct(struct_env, self.type_inst, false);
-
         // Emit $UpdateField functions.
         let fields = struct_env.get_fields().collect_vec();
-        for (pos, field_env) in fields.iter().enumerate() {
-            let field_name = field_env.get_name().display(env.symbol_pool()).to_string();
-            self.emit_function(
-                &format!(
-                    "$Update'{}'_{}(s: {}, x: {}): {}",
-                    suffix,
-                    field_name,
-                    struct_name,
-                    self.boogie_type_for_struct_field(
+
+        if struct_env.has_variants() {
+            let mut field_variant_map: BTreeMap<(String, String), Vec<String>> = BTreeMap::new();
```

Comment:
Maybe worse to factor this out into independent function.

Reply:
done